src/lib/cache/gdfs_cache.hpp

```diff
@@ -105,53 +93,60 @@ class GDFSCache : public AbstractCacheImpl<Key, Value> {
     return entry.value;
   }
 
-  bool has(const Key& key) const { return _map.find(key) != _map.end(); }
+  bool has(const Key& key) const {
+    std::shared_lock<std::shared_mutex> lock(_mutex);
+    return _map.find(key) != _map.end();
+  }
 
-  size_t size() const { return _map.size(); }
+  size_t size() const {
+    std::shared_lock<std::shared_mutex> lock(_mutex);
+    return _map.size();
+  }
 
   void clear() {
+    std::unique_lock<std::shared_mutex> lock(_mutex);
     _map.clear();
     _queue.clear();
   }
 
   void resize(size_t capacity) {
+    std::unique_lock<std::shared_mutex> lock(_mutex);
     while (_queue.size() > capacity) {
       _evict();
     }
     this->_capacity = capacity;
   }
 
-  const boost::heap::fibonacci_heap<GDFSCacheEntry>& queue() const { return _queue; }
-
-  double inflation() const { return _inflation; }
-
-  double priority(const Key& key) const {
-    auto it = _map.find(key);
-    return (*it->second).priority;
-  }
-
-  ErasedIterator begin() { return ErasedIterator{std::make_unique<Iterator>(_map.begin())}; }
-
-  ErasedIterator end() { return ErasedIterator{std::make_unique<Iterator>(_map.end())}; }
-
-  size_t frequency(const Key& key) {
+  size_t frequency(const Key& key) const {
+    std::shared_lock<std::shared_mutex> lock(_mutex);
     const auto it = _map.find(key);
     if (it == _map.end()) {
       return size_t{0};
     }
+    return (*it->second).frequency;
+  }
 
-    Handle handle = it->second;
-    GDFSCacheEntry& entry = (*handle);
-    return entry.frequency;
+  std::unordered_map<Key, SnapshotEntry> snapshot() const {
+    std::shared_lock<std::shared_mutex> lock(_mutex);
+    std::unordered_map<Key, SnapshotEntry> map_copy(_map.size());
+    for (const auto& [key, entry] : _map) {
+      map_copy[key] = SnapshotEntry{(*entry).value, (*entry).frequency};
+    }
+    return map_copy;
   }
 
  protected:
+  friend class CachePolicyTest;
+
   // Priority queue to hold all elements. Implemented as max-heap.
   boost::heap::fibonacci_heap<GDFSCacheEntry> _queue;
 
   // Map to point towards element in the list.
+  // We use a locked map here since TBB's concurrent maps are not safe for erasing.
```

Comment:
Isn't this also an issue of the heap and the map having to be synchronized?

Suggestion:
* Make this a comment for the _mutex, not for _map. 
* "Locking this data structure is easier than using a concurrent data structure (e.g. TBB) as (1) both the queue and the map would have to be concurrent data structures, (2) their modifications would have to be synchronized, and (3) TBB's concurrent_unordered_map does not provide for deletion and concurrent_hash_map is difficult to use."